File: scripts/process_raw_data.py
import argparse
import numpy as np
import os
import pandas as pd
import time
from drms import Client

parser = argparse.ArgumentParser(description="Process raw HARP data downloaded from JSOC")
parser.add_argument("--series", type=str, required=True, help="A data series for which raw data has been downloaded, e.g., hmi.sharp_cea_720s")
parser.add_argument("--keep_low_qual_vals", action="store_true", help="Whether to keep predictor values in low-quality records (default: False)")
parser.add_argument("--keep_near_limb_recs", action="store_true", help="Whether to keep records for near-limb HARPs (default: False)")
parser.add_argument("--limb_threshold", type=float, default=70, help="Longitude threshold defining where the limb starts (default: 70)")
cmd_args = parser.parse_args()
series = cmd_args.series
series_no_dot = series.replace(".", "_", 1)
keep_low_qual_vals = cmd_args.keep_low_qual_vals
keep_near_limb_recs = cmd_args.keep_near_limb_recs
limb_threshold = cmd_args.limb_threshold

print("Compiling the raw data", end="", flush=True)
start_time = time.time()
files = [os.path.join("raw", file) for file in os.listdir("raw") if file.endswith(".parquet")]
files.sort()
data = pd.concat([pd.read_parquet(file) for file in files], ignore_index=True)
elapsed_time = time.time() - start_time
print(f"\rCompiling the raw data ({int(elapsed_time)}s)", flush=True)

print("Turning datetime strings into datetimes", end="", flush=True)
start_time = time.time()
cli = Client()
keywords = cli.info(series).keywords
types = keywords[keywords.index.isin(data.columns)]["type"]
time_cols = types[types == "time"].index
data[time_cols] = data[time_cols].apply(
    # The TAI system differs from the UTC system; see (1). The latter system is used in the GOES data (2, 3). However, the discrepancy is only 37
    # seconds as of 2025-01-11, so the impact of treating TAI times as UTC times should be negligible. It seems to be possible to convert TAI
    # times to UTC times using the astropy package though (4).
    #
    # (1) https://en.wikipedia.org/wiki/International_Atomic_Time
    # (2) https://www.ncei.noaa.gov/data/goes-space-environment-monitor/access/science/xrs/GOES_1-15_XRS_Science-Quality_Data_Readme.pdf
    # (3) https://data.ngdc.noaa.gov/platforms/solar-space-observing-satellites/goes/goes16/l2/docs/GOES-R_XRS_L2_Data_Users_Guide.pdf
    # (4) https://docs.astropy.org/en/stable/time/index.html
    lambda col: pd.to_datetime(col.str.replace("_TAI", ""), utc=True, format="%Y.%m.%d_%H:%M:%S")
)
elapsed_time = time.time() - start_time
print(f"\rTurning datetime strings into datetimes ({int(elapsed_time)}s)", flush=True)

# In a tiny number of records, some variables have infinite values. For example, for HARP 3672, at 2014-01-21 12:48:00, ABSNJZH is infinite.
# These infinities are probably invalid values, so we replace them with NaNs.
print("Replacing infinities with NaNs", end="", flush=True)
start_time = time.time()
number_cols = data.select_dtypes(include=["number"]).columns
data[number_cols] = data[number_cols].replace([np.inf, -np.inf], np.nan)
elapsed_time = time.time() - start_time
print(f"\rReplacing infinities with NaNs ({int(elapsed_time)}s)", flush=True)

predictors = [
    # SHARP parameters
    "USFLUX", "MEANGAM", "MEANGBT", "MEANGBZ", "MEANGBH", "MEANJZD", "TOTUSJZ", "MEANALP",
    "MEANJZH", "TOTUSJH", "ABSNJZH", "SAVNCPP", "MEANPOT", "TOTPOT", "MEANSHR", "SHRGT45",
    # Patch areas and pixel counts
    "NPIX", "SIZE", "AREA", "NACR", "SIZE_ACR", "AREA_ACR"
]
if not keep_low_qual_vals:
    if (data.columns == "QUALITY").any():
        print("Replacing low-quality values with NaNs", end="", flush=True)
        start_time = time.time()
        data.loc[data["QUALITY"] != 0, predictors] = np.nan
        elapsed_time = time.time() - start_time
        print(f"\rReplacing low-quality values with NaNs ({int(elapsed_time)}s)", flush=True)
    else:
        raise ValueError("To replace low-quality values with NaNs, QUALITY must be in the list of keywords")

# Per (1), the level of noise in the SHARP parameters is substantially higher when the HARP is near the limb of the Sun. Per (3)-(5),
# data measured on HARPs near the limb can be distorted by projection effects. Reference (1) seems to define the limb as the region
# more than 45 degrees from the central meridian (see Section 9), while reference (2) uses a threshold of 70 degrees. However, in
# (2), no X-class flare in the data occurred more than 68 degrees from the central meridian. References (3)-(5) use a threshold of 68
# degrees, seemingly because of (2). Based on Table 4 in (1) and (5), the longitude of the flux-weighted center of the active pixels
# (LON_FWT) is what needs to be within bounds.
#
# (1) https://doi.org/10.1007/s11207-014-0529-3
# (2) https://dx.doi.org/10.1088/0004-637X/798/2/135
# (3) https://doi.org/10.1029/2019SW002214
# (4) https://doi.org/10.1029/2020SW002440
# (5) https://doi.org/10.3847/1538-4357/ab89ac
if not keep_near_limb_recs:
    if (data.columns == "LON_FWT").any():
        print("Discarding near-limb records", end="", flush=True)
        start_time = time.time()
        data = data[data["LON_FWT"].abs() <= limb_threshold]
        elapsed_time = time.time() - start_time
        print(f"\rDiscarding near-limb records ({int(elapsed_time)}s)", flush=True)
    else:
        raise ValueError("To discard near-limb records, LON_FWT must be in the list of keywords")

def insert_missing_rows(group: pd.DataFrame) -> pd.DataFrame:
    """
    Insert rows in the Pandas DataFrame for a specified HARP for times that should have rows but do not.

    :param group: Pandas DataFrame containing the data for the HARP.
    :return: Pandas DataFrame with the missing rows added.
    """
    full_time_range = pd.date_range(start=group["T_REC"].min(), end=group["T_REC"].max(), freq="12min")
    group = group.set_index("T_REC").reindex(full_time_range, fill_value=pd.NA).reset_index(names="T_REC")
    group.insert(1, "T_REC", group.pop("T_REC"))
    group["HARPNUM"] = group["HARPNUM"].ffill().bfill()
    return group

print("Inserting missing rows", end="", flush=True)
start_time = time.time()
orig_dtypes = data.dtypes
orig_dtypes.where(orig_dtypes != np.dtype("int64"), pd.Int64Dtype(), inplace=True) # int64 doesn't support NA, but Int64 does
data = data.groupby("HARPNUM")[data.columns].apply(insert_missing_rows)
data.reset_index(drop=True, inplace=True)
data = data.astype(orig_dtypes)
data.sort_values(by=["HARPNUM", "T_REC"], inplace=True)
elapsed_time = time.time() - start_time
print(f"\rInserting missing rows ({int(elapsed_time)}s)", flush=True)

print("Saving the DataFrame", end="", flush=True)
start_time = time.time()
if keep_low_qual_vals and keep_near_limb_recs:
    file_name = "all"
else:
    high_qual_str = "" if keep_low_qual_vals else "hq"
    limb_threshold = int(limb_threshold) \
        if limb_threshold.is_integer() else limb_threshold
    near_center_str = "" if keep_near_limb_recs else f"nc{limb_threshold}"
    file_name = "_".join(filter(None, [high_qual_str, near_center_str]))
data.to_parquet(f"processed/{file_name}.parquet")
elapsed_time = time.time() - start_time
print(f"\rSaving the DataFrame ({int(elapsed_time)}s)", flush=True)
print("Done")

File: scripts/R/process_raw_data.R
# here::i_am("../../scripts/process_raw_harp_data.R")
setwd("~/research/harp_data/")

# Load packages -----------------------------------------------------------

library(foreach)
# library(here)
library(lubridate)
library(reticulate)
library(tidyverse)
library(tools)
use_condaenv("env/")
# use_condaenv(here("conda/env"))
drms <- import("drms")

# Process the data --------------------------------------------------------

# The command line arguments should be
# - the file with the keywords
# - the subdirectory where downloaded data has been saved
# - the RData file the processed data should be put in,
# in that order. The keywords file should be the R script in data/raw/harp/ that
# creates a character vector called keywords containing the keywords. The
# subdirectory is assumed to be in data/raw/harp/. The RData file is created in
# data/processed/harp/. If the current script is being run interactively, it
# should be sourced so that the scan() call works properly
if (interactive()) {
  cat("\nEnter keywords file, subdirectory with data files, and output file\n")
  command_args <- scan(what = character(), nmax = 3)
} else {
  command_args <- commandArgs(trailingOnly = TRUE)
}
stopifnot(length(command_args) == 3)
keywords_file <- command_args[1]
subdir <- command_args[2]
output_file <- command_args[3]

source(here(str_glue("data/raw/harp/{keywords_file}")))
# See the comments in data/raw/harp/download_data.R for information on the use
# of this object
cli <- drms$Client()
spec_expr <- cli$info("hmi.sharp_720s")$keywords %>%
  as_tibble(rownames = "keyword") %>%
  select(keyword, type) %>%
  # HARPNUM and T_REC are primary keys of hmi.sharp_720s as can be verified
  # using the code in the comments referred to above
  filter(keyword %in% c("HARPNUM", "T_REC", keywords)) %>%
  mutate(
    spec_fun = case_when( # This may need to be updated over time
      type %in% c("double", "float", "longlong") ~ "col_double()",
      type == "int" ~ "col_integer()",
      type %in% c("string", "time") ~ "col_character()"
    ),
    tag_val_pair = str_c(keyword, spec_fun, sep = " = ")
  ) %>%
  pull(tag_val_pair) %>%
  str_c(collapse = ",") %>%
  {str_glue("spec <- cols({.})")} %>%
  parse(text = .)
eval(spec_expr)

files <- list.files(here(str_glue("data/raw/harp/{subdir}")), full.names = TRUE)
harp_tbl <- foreach (file = files, .combine = bind_rows) %do% {
  writeLines(file_path_sans_ext(basename(file)))
  read_csv(file, col_types = spec, show_col_types = FALSE)
} %>%
  mutate(
    T_REC = parse_date_time(str_remove(T_REC, "_TAI"), "%Y.%m.%d_%H:%M:%S"),
    # See http://jsoc.stanford.edu/jsocwiki/Lev1qualBits for information about
    # this keyword
    QUALITY = str_c("0x", str_pad(as.hexmode(QUALITY), width = 8, pad = "0")),
    QUAL_S = str_c("0x", str_pad(as.hexmode(QUAL_S), width = 8, pad = "0")),
    # We flag records for HARPs that were entirely within 68 degrees of the
    # central meridian at the time of observation. Using only these records is
    # a way to avoid projection effects as stated in (1) and (2).
    #
    # (1) https://doi.org/10.1029/2020SW002440
    # (2) https://doi.org/10.3847/1538-4357/ab89ac
    are_lons_small = (abs(LON_MIN) <= 68) & (abs(LON_MAX) <= 68)
  ) %>%
  nest(.by = HARPNUM) %>%
  # There should be a set of SHARP parameter values every 12 minutes, but that
  # seems to be false; see eda/missingness_analysis/missingness_analysis.Rmd
  mutate(
    data = map(data, ~ complete(.x, T_REC = full_seq(T_REC, period = 720)))
  ) %>%
  unnest(data)

# This will come in handy in several places
sharp_params <- c(
  "ABSNJZH", "MEANALP", "MEANGAM", "MEANGBH", "MEANGBT", "MEANGBZ", "MEANJZD",
  "MEANJZH", "MEANPOT", "MEANSHR", "SAVNCPP", "SHRGT45", "TOTPOT", "TOTUSJH",
  "TOTUSJZ", "USFLUX"
)

# Save the processed data -------------------------------------------------

save(
  harp_tbl, sharp_params,
  file = here(str_glue("data/processed/harp/{output_file}"))
)